File: linked-list/linked-list.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None
        self.prev = None

class DoublyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None

    def add_to_front(self, value):
        new_node = Node(value)
        if not self.head:
            self.head = self.tail = new_node
        else:
            new_node.next = self.head
            self.head.prev = new_node
            self.head = new_node

    def add_to_end(self, value):
        new_node = Node(value)
        if not self.tail:
            self.head = self.tail = new_node
        else:
            new_node.prev = self.tail
            self.tail.next = new_node
            self.tail = new_node

    def remove_from_front(self):
        if not self.head:
            return None
        removed_value = self.head.value
        if self.head == self.tail:
            self.head = self.tail = None
        else:
            self.head = self.head.next
            self.head.prev = None
        return removed_value

    def remove_from_end(self):
        if not self.tail:
            return None
        removed_value = self.tail.value
        if self.head == self.tail:
            self.head = self.tail = None
        else:
            self.tail = self.tail.prev
            self.tail.next = None
        return removed_value
    
dll = DoublyLinkedList()

print(dll.remove_from_front())  
print(dll.remove_from_end())    

dll.add_to_front(100)
print(dll.head.value) 
print(dll.tail.value) 

dll.add_to_end(200)
print(dll.head.value) 
print(dll.tail.value) 

dll.add_to_front(50)
print(dll.head.value)  
print(dll.tail.value)  

removido = dll.remove_from_front()
print(removido)        
print(dll.head.value)  
print(dll.tail.value)  

removido = dll.remove_from_end()
print(removido)        
print(dll.head.value)  
print(dll.tail.value)  

removido = dll.remove_from_front()
print(removido)        
print(dll.head)        
print(dll.tail)        

dll.add_to_end(1)
dll.add_to_end(2)
dll.add_to_end(3)
print(dll.head.value)  
print(dll.tail.value)  

print(dll.remove_from_front())  
print(dll.remove_from_end())    
print(dll.remove_from_front())  

print(dll.head)  
print(dll.tail)  

File: linked-list/linked-list.js
class Node {
  constructor(value) {
    this.value = value // value stored in the node
    this.next = null // pointer to the next node
    this.prev = null // pointer to the previous node
  }
}

class DoublyLinkedList {
  constructor() {
    this.head = null // beginning of the list
    this.tail = null // end of the list
  }

  addToFront(value) {
    const newNode = new Node(value)

    if (!this.head) {
      this.head = this.tail = newNode
    } else {
      newNode.next = this.head
      this.head.prev = newNode
      this.head = newNode
    }
  }

  addToEnd(value) {
    const newNode = new Node(value)

    if (!this.tail) {
      this.head = this.tail = newNode
    } else {
      newNode.prev = this.tail
      this.tail.next = newNode
      this.tail = newNode
    }
  }

  removeFromFront() {
    if (!this.head) return null

    const removedValue = this.head.value

    if (this.head === this.tail) {
      this.head = this.tail = null
    } else {
      this.head = this.head.next
      this.head.prev = null
    }

    return removedValue
  }

  removeFromEnd() {
    if (!this.tail) return null
    const removedValue = this.tail.value

    if (this.head === this.tail) {
      this.head = this.tail = null
    } else {
      this.tail = this.tail.prev
      this.tail.next = null
    }

    return removedValue
  }
}

const list = new DoublyLinkedList()


console.log(list.removeFromFront()) 
console.log(list.removeFromEnd())   

list.addToFront(10)
console.log(list.head.value)        
console.log(list.tail.value)        

list.addToFront(20)
console.log(list.head.value)        
console.log(list.head.next.value)   
console.log(list.tail.value)        
console.log(list.tail.prev.value)   

list.addToEnd(30)
console.log(list.tail.value)        
console.log(list.tail.prev.value)   
console.log(list.head.value)        

console.log(list.removeFromFront()) 
console.log(list.head.value)        
console.log(list.tail.value)        

console.log(list.removeFromEnd())   
console.log(list.tail.value)        
console.log(list.head.value)        

console.log(list.removeFromFront()) 
console.log(list.head)              
console.log(list.tail)              

list.addToEnd(100)
list.addToEnd(200)
list.addToEnd(300)
console.log(list.head.value)        
console.log(list.tail.value)        
console.log(list.head.next.value)   
console.log(list.tail.prev.value)   

console.log(list.removeFromEnd())   
console.log(list.removeFromFront()) 
console.log(list.removeFromFront()) 
console.log(list.head)              
console.log(list.tail)              
